indexing-service/src/main/java/io/druid/indexing/common/task/ArchiveTask.java

```diff
@@ -93,12 +101,41 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       log.info("OK to archive segment: %s", unusedSegment.getIdentifier());
     }
 
-    // Move segments
+
+    // Archiving segments can take time which seems to be a lot of work to do in critical section
+    // Thus, after each batch downgrade the lock to ensure liveliness of the system
+    // so that higher priority tasks are not blocked for a long time
+    // Note - If we are unable to upgrade the lock again then the task will fail,
+    // there is no point in retrying here as upgrade should fail only when the taskLock is removed by a higher priority task
+    int counter = 0;
     for (DataSegment segment : unusedSegments) {
+      if (counter % getBatchSize() == 0) {
```

Comment:
It would be good if downgrade-upgrade rotation was driven by time elasped since last rotation rather than number of segments.